File: solutions/csharp/log-levels/1/LogLevels.cs
using System;

static class LogLine
{
    public static string Message(string logLine)
    {
        var levelEnd = logLine.IndexOf(":");
        var msgStart = levelEnd + 1;
        if(levelEnd >= 0 && msgStart < logLine.Length){
            return logLine.Substring(msgStart).Trim();
        }
        return logLine;
    }

    public static string LogLevel(string logLine)
    {
        var levelStart = logLine.IndexOf("[");
        var levelEnd = logLine.IndexOf("]");
        if(levelStart >= 0 && levelStart + 1 < logLine.Length && levelEnd >= 0 && levelEnd < logLine.Length){
            return logLine.Substring(levelStart + 1, levelEnd - levelStart-1).ToLower();
        }
        return logLine;
    }

    public static string Reformat(string logLine)
    {
        return Message(logLine) + $" ({LogLevel(logLine)})";
    }
}

File: solutions/cpp/log-levels/1/log_levels.cpp
#include <string>
using std::string;

namespace log_line {
string message(std::string line) {
  int msg_start = line.find(':') + 2;
  return line.substr(msg_start);
}

string log_level(std::string line) {
  int log_level_start = line.find('[') + 1;
  int log_level_end = line.find(']');
  return line.substr(log_level_start, log_level_end - log_level_start);
}

string reformat(std::string line) {
  return message(line) + " (" + log_level(line) + ")";
}
} // namespace log_line
